src/v/resource_mgmt/storage.cc

```diff
@@ -218,6 +218,141 @@ eviction_policy::partition* eviction_policy::schedule::current() {
+        if (!p.second->remote_partition()) {
+            continue;
+        }
+        partitions.push_back(p.second);
+    }
+
+    /*
+     * retention settings mirror settings found in housekeeping()
+     */
+    const auto collection_threshold = [this] {
+        const auto& lm = _storage->local().log_mgr();
+        if (!lm.config().delete_retention().has_value()) {
+            return model::timestamp(0);
+        }
+        const auto now = model::timestamp::now().value();
+        const auto retention = lm.config().delete_retention().value().count();
+        return model::timestamp(now - retention);
+    };
+
+    gc_config cfg(
+      collection_threshold(),
+      _storage->local().log_mgr().config().retention_bytes());
+
+    /*
+     * in smallish batches partitions are queried for their reclaimable
+     * segments. all of this information is bundled up and returned.
+     */
+    fragmented_vector<partition> res;
+    co_await ss::max_concurrent_for_each(
+      partitions.begin(), partitions.end(), 20, [&res, cfg](const auto& p) {
+          auto log = dynamic_cast<storage::disk_log_impl*>(p->log().get_impl());
+          auto gate = log->gate().hold(); // protect against log deletes
+          return log->get_reclaimable_offsets(cfg)
+            .then([&res, p](auto offsets) {
+                res.push_back({
+                  .group = p->group(),
+                  .offsets = std::move(offsets),
+                });
+            })
+            .finally([g = std::move(gate)] {});
+      });
+
+    /*
+     * sorting by raft::group_id would provide a more stable round-robin
+     * evaluation of partitions in later processing of the result, but on small
+     * timescales we don't expect enough churn to make a difference, nor with a
+     * large number of partitions on the system would the non-determinism
+     * in the parallelism above in max_concurrent_for_each be problematic.
+     */
+
+    vlog(rlog.trace, "Reporting reclaim offsets for {} partitions", res.size());
+    co_return res;
+}
+
+ss::future<> eviction_policy::install_schedule(schedule schedule) {
+    /*
+     * install the schedule on each core in parallel
+     */
+    auto total = co_await ss::map_reduce(
+      schedule.shards,
+      [this](auto& shard) { return install_schedule(std::move(shard)); },
+      size_t(0),
+      std::plus<>());
+
+    vlog(rlog.info, "Requested truncation from {} cloud partitions", total);
+}
+
+// the per-shard counterpart of install_schedule(schedule)
+ss::future<size_t> eviction_policy::install_schedule(shard_partitions shard) {
+    const auto shard_id = shard.shard;
+    return _pm->invoke_on(shard_id, [shard = std::move(shard)](auto& pm) {
+        size_t decisions = 0;
+        for (const auto& partition : shard.partitions) {
+            if (!partition.decision.has_value()) {
+                continue;
+            }
+
+            auto p = pm.partition_for(partition.group);
```

Comment:
Should we somehow communicate to the caller that the partition was moved? Perhaps in this case it could use metadata to find partition's shard and try to install schedule again.

Reply:
i'm not really sure it is worth the added complexity. once we install the schedule the space management will continue running until we meet the goal. if we didn't meet the goal because some partition got moved off the node, then we'd find a replacement the next time around. wdyt?